File: Python/clbg/pidigits/pidigits.py
import sys


def main():
    target = int(sys.argv[1])

    q, r, t = 1, 0, 1
    k, digit, l = 1, 3, 3

    produced = 0
    group = []
    output = []

    while produced < target:
        digit_t = digit * t

        if (q << 2) + r - t < digit_t:
            group.append(chr(48 + digit))
            produced += 1

            old_q, old_r = q, r
            q = old_q * 10
            r = (old_r - digit_t) * 10
            digit = (10 * (3 * old_q + old_r)) // t - digit * 10

            if len(group) == 10 or produced == target:
                digits = ''.join(group)
                output.append(digits.ljust(10) + "\t:" + str(produced))
                group.clear()
        else:
            qk = q * k
            rl = r * l
            tl = t * l

            digit = (7 * qk + rl + 2) // tl
            r = rl + (qk << 2) + (q << 1)
            q = qk
            t = tl
            k += 1
            l += 2

    sys.stdout.write('\n'.join(output) + '\n')


if __name__ == "__main__":
    main()

File: TypeScript/clbg/pidigits/pidigits.js
const limit = Number.parseInt(process.argv[2], 10);
let q = 1n;
let r = 0n;
let t = 1n;
let k = 1n;
let digit = 3n;
let l = 3n;
let produced = 0;
let group = "";
const lines = [];
while (produced < limit) {
    if (4n * q + r - t < digit * t) {
        group += digit.toString();
        produced++;
        if (group.length === 10) {
            lines.push(`${group}\t:${produced}`);
            group = "";
        }
        const nextR = 10n * (r - digit * t);
        digit = (10n * (3n * q + r)) / t - 10n * digit;
        q *= 10n;
        r = nextR;
    }
    else {
        const nextR = (2n * q + r) * l;
        const nextT = t * l;
        const nextDigit = (q * (7n * k) + 2n + r * l) / nextT;
        q *= k;
        t = nextT;
        r = nextR;
        k += 1n;
        l += 2n;
        digit = nextDigit;
    }
}
if (group.length > 0) {
    lines.push(`${group.padEnd(10, " ")}\t:${produced}`);
}
process.stdout.write(lines.join("\n") + "\n");
